sycl/include/CL/sycl/INTEL/esimd/esimd.hpp

```diff
@@ -179,9 +179,29 @@ template <typename Ty, int N> class simd {
   //   This would allow you to use the subscript operator to write to an
   //   element.
   // {/quote}
-  /// Read a single element, by value only.
+  /// Read single element, return value only (not reference).
   Ty operator[](int i) const { return data()[i]; }
 
+  /// Read multiple elements by their indices in vector
+  template <int Size> simd<Ty, Size> iselect(simd<uint16_t, Size> &Indices) {
+    vector_type_t<uint16_t, Size> Offsets = Indices.data() * sizeof(Ty);
+    return __esimd_rdindirect<Ty, N, Size>(data(), Offsets);
+  }
+  /// update single element
+  void iupdate(ushort Index, Ty V) {
+    auto Val = data();
+    Val[Index] = V;
+    set(Val);
+  }
+  /// update multiple elements by their indices in vector
+  template <int Size>
+  void iupdate(simd<uint16_t, Size> &Indices, simd<Ty, Size> &Val,
+               mask_type_t<Size> Mask) {
+    vector_type_t<uint16_t, Size> Offsets = Indices.data() * sizeof(Ty);
+    M_data = __esimd_wrindirect<Ty, N, Size>(data(), Val.data(), Offsets, Mask);
+    set(M_data);
```

Comment:
This is redundant for setting M_data? Or do we need a temp to perform v_store to M_data?